### ISSUE TITLE
Upwards arrow on empty buffer should bring last message into edit mode

### ISSUE THREAD

**Issue by Tmpod** [cinnyapp/cinny#97]
#### Is your feature request related to a problem? Please describe.

Editing the last message sent to quickly fix a mistake is a bit of a hassle at the moment. I have to grab my mouse and click the edit button.

#### Describe the solution you'd like

Like in many other clients and platforms (like Discord, which you're trying to replicate) allow you to enter edit mode on that last message by clicking on the upwards arrow when the message buffer is empty. This should be implemented in Cinny.

#### Describe alternatives you've considered

Some other key could be used, but I haven't thought of anything else outside of that.

**Comment by blattersturm** [comment]
In addition to this, I noticed these two flow things:

1. Manually clicking the edit button doesn't seem to focus the edit field, requiring a second click. This is probably best off as a second issue, however.
2. In Element, pressing up/down when at the start or end of an edit text box seems to move up to editing the message above/below it. This feature could be useful?

**Comment by Tmpod** [comment]
1. Yeah, I noticed that too. Best off as a separate issue like you suggested.
2. Yup, that would be neat to have. So if you wanted to edit the second to last message you could just quickly double tap the up arrow key.

**Comment by Litom** [comment]
> #### Describe alternatives you've considered
> 
> Some other key could be used, but I haven't thought of anything else outside of that.

I think the up arrow key is already a good standard for this. At most, it would be interesting to be able to change these keys in the settings according to one's own preferences (along with future other shortcut options). Perhaps in a separate menu tab called e.g., "Keybindings".

**Comment by Tmpod** [comment]
Yeah, it's definitely a good idea to make it customisable!

**Comment by BachoSeven** [comment]
This would go a long way of making Cinny more accessible to heavy keyboard users.

Just wanted to add that when this gets implemented it would be nice to also have `Ctrl+Up/DownArrow` to reply to messages, as it works in element.

**Comment by deanveloper** [comment]
the way that slack and discord both work (which i really like)

empty text box: up-arrow edits previous message
anything in text box: up-arrow moves caret, even if it's at the beginning of the text box it will not ever edit

another feature i really like that ties into this - pressing tab highlights the "send message" box, and using the up/down arrows navigate through messages. pressing `E` will edit the currently highlighted message, allowing you to edit messages besides the last one you sent

**Comment by exodrifter** [comment]
I talked about this with a friend and we would like to throw our hat in the ring. We think it would be better if pressing up highlighted the last message that was sent in the room and allowed you to pick between any of the message options you normally get on mouse hover.

This way, you could press up arrow until you get to the message you want, use the left and right arrow buttons to navigate between message actions, and use the enter key to do do that action. If we designed it this way, it would be nice to automatically select the edit button and place it in the middle so you would only need one keypress to pick either one of the other two options.

This solves a few problems:
* Pressing up while you're editing a message to get to your next-to-last message seems undesirable since you could potentially lose edits.
* You don't need to add more key combinations do other actions (such as the aforementioned pressing tab _then_ up and down to select messages).

**Comment by blattersturm** [comment]
> use the left and right arrow buttons to navigate between message actions, and use the enter key to do do that action

In this case, would it make sense to also add a set of keys (similar to what Zulip's UX does here) for common actions - e.g. \> for reply, e for edit, and so on, _in addition_ to the arrow-based selection?

**Comment by ajbura** [comment]
> pressing up highlighted the last message that was sent in the room

@exodrifter Will it highlight all messages or just the one sent by user?

**Comment by deanveloper** [comment]
I don't want to put words in their mouth, but I believe all messages. This can useful for other keybinds, such as R for replying, and D for deleting messages (since mods/admins can delete other people's messages).

**Comment by exodrifter** [comment]
@ajbura Yes, we were imagining that it would cycle through all messages and not just your own, so you could use other keybindings as @deanveloper correctly assumed.

**Comment by Tmpod** [comment]
We could also have a distinction between "Up" and "Shift+Up" -- one could go
through every message and the other would skip others' messages.

**Comment by ajbura** [comment]
I think we are mixing two use cases here: First to quickly edit last sent message and second general keyboard navigation in timeline. 

For editing the last sent message, Up arrow seems to be most commonly used and most app only allow to edit only last message, which make sense as most of the time people only want to edit just last message.

For the timeline navigation, the timeline scroll with up/down arrow works already and we can polish that to support reply, delete etc.

**Comment by Tmpod** [comment]
> I think we are mixing two use cases here: First to quickly edit last sent
> message and second general keyboard navigation in timeline. 

I'd say we could not treat quickly editing your last message as a special
case and just have it as a timeline action. `Up``e` really isn't harder than
just `Up`, imo.
